File: app/src/main/java/com/orgzly/android/ui/util/ActivityUtils.java
package com.orgzly.android.ui.util;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.res.TypedArray;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import com.orgzly.BuildConfig;
import com.orgzly.R;
import com.orgzly.android.ui.MainActivity;
import com.orgzly.android.ui.fragments.BookPrefaceFragment;
import com.orgzly.android.ui.fragments.BookFragment;
import com.orgzly.android.ui.fragments.BooksFragment;
import com.orgzly.android.ui.fragments.FilterFragment;
import com.orgzly.android.ui.fragments.FiltersFragment;
import com.orgzly.android.ui.fragments.NoteFragment;
import com.orgzly.android.ui.fragments.QueryFragment;
import com.orgzly.android.ui.fragments.ReposFragment;
import com.orgzly.android.ui.fragments.SettingsFragment;
import com.orgzly.android.util.LogUtils;

/**
 *
 */
public class ActivityUtils {
    private static final String TAG = ActivityUtils.class.getName();

    public static void closeSoftKeyboard(final Activity activity) {
        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, "Hiding keyboard in activity " + activity);

        View view = activity.getCurrentFocus();

        /* If no view currently has focus, create a new one to grab a window token from it. */
        if (view == null) {
            if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, "No view in focus, using activity");
            view = new View(activity);
        }

        final View finalView = view;

//        new Handler().postDelayed(new Runnable() {
//            @Override
//            public void run() {
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(finalView.getWindowToken(), 0);
//            }
//        }, 100);
    }

    public static void openSoftKeyboard(final Activity activity, final View view) {
        if (view.requestFocus()) {
            if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, "Showing keyboard for view " + view + " in activity " + activity);

            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    InputMethodManager imm = (InputMethodManager) activity.getSystemService(Context.INPUT_METHOD_SERVICE);
                    imm.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT);
                }
            }, 100);

        } else {
            Log.w(TAG, "Can't open keyboard because view " + view +
                       " failed to get focus in activity " + activity);
        }
    }

    /**
     * Color the action bar depending on which fragment is displayed.
     */
    public static void setColorsForFragment(final AppCompatActivity activity, String fragmentTag) {
        FragmentResources resources = new FragmentResources(activity, fragmentTag);

        /* Color status bar. */
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            activity.getWindow().setStatusBarColor(resources.statusColor);
            // getWindow().setNavigationBarColor(color);
        }

        /* Color action bar. */
        activity.getSupportActionBar()
                .setBackgroundDrawable(new ColorDrawable(resources.actionColor));
    }

    public static class FragmentResources {
        public int statusColor;
        public int actionColor;
        public Drawable fabDrawable;

        public FragmentResources(Context context, String fragmentTag) {
            int statusBarAttr;
            int actionBarAttr;
            int fabAttr;

            if (SettingsFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_in_settings;
                actionBarAttr = R.attr.action_bar_in_settings;
                fabAttr = 0;

            } else if (ReposFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_in_settings;
                actionBarAttr = R.attr.action_bar_in_settings;
                fabAttr = 0;

            } else if (FiltersFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_in_query;
                actionBarAttr = R.attr.action_bar_in_query;
                fabAttr = R.attr.oic_new_item;

            } else if (FilterFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_in_query;
                actionBarAttr = R.attr.action_bar_in_query;
                fabAttr = 0;

            } else if (QueryFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_in_query;
                actionBarAttr = R.attr.action_bar_in_query;
                fabAttr = 0;

            } else if (BooksFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_default;
                actionBarAttr = R.attr.action_bar_default;
                fabAttr = R.attr.oic_new_item;

            } else if (BookFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_default;
                actionBarAttr = R.attr.action_bar_default;
                fabAttr = R.attr.oic_new_item;

            } else if (BookPrefaceFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_default;
                actionBarAttr = R.attr.action_bar_default;
                fabAttr = 0;

            } else if (NoteFragment.FRAGMENT_TAG.equals(fragmentTag)) {
                statusBarAttr = R.attr.status_bar_default;
                actionBarAttr = R.attr.action_bar_default;
                fabAttr = 0;

            } else {
                statusBarAttr = R.attr.status_bar_default;
                actionBarAttr = R.attr.action_bar_default;
                fabAttr = 0;
            }

            TypedArray typedArray = context.obtainStyledAttributes(new int[] {
                    statusBarAttr,
                    actionBarAttr,
                    fabAttr
            });

            statusColor = typedArray.getColor(0, 0);
            actionColor = typedArray.getColor(1, 0);
            fabDrawable = typedArray.getDrawable(2);

            typedArray.recycle();
        }
    }

    /**
     * Open "App info" settings, where permissions can be granted.
     */
    public static void openAppInfoSettings(Activity activity) {
        Intent intent = new Intent(android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS);

        intent.setData(Uri.parse("package:" + BuildConfig.APPLICATION_ID));

        activity.startActivity(intent);
    }

    public static PendingIntent mainActivityPendingIntent(Context context, long bookId, long noteId) {
        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, bookId, noteId);
        Intent intent = Intent.makeRestartActivityTask(new ComponentName(context, MainActivity.class));

        intent.putExtra(MainActivity.EXTRA_BOOK_ID, bookId);
        intent.putExtra(MainActivity.EXTRA_NOTE_ID, noteId);

        return PendingIntent.getActivity(
                context,
                Long.valueOf(noteId).intValue(),
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT);
    }
}

File: app/src/main/java/com/orgzly/android/ui/util/ActivityUtils.kt
package com.orgzly.android.ui.util

import android.app.Activity
import android.app.PendingIntent
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.net.Uri
import android.os.Build
import android.util.DisplayMetrics
import android.view.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.view.menu.ActionMenuItemView
import androidx.appcompat.widget.Toolbar
import androidx.core.view.WindowCompat
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.orgzly.BuildConfig
import com.orgzly.R
import com.orgzly.android.AppIntent
import com.orgzly.android.prefs.AppPreferences
import com.orgzly.android.ui.main.MainActivity
import com.orgzly.android.util.LogUtils

object ActivityUtils {
    private val TAG = ActivityUtils::class.java.name

    /**
     * Open "App info" settings, where permissions can be granted.
     */
    fun openAppInfoSettings(activity: Activity) {
        val intent = Intent(android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS)

        intent.data = Uri.parse("package:" + BuildConfig.APPLICATION_ID)

        activity.startActivity(intent)
    }

    @JvmStatic
    fun mainActivityPendingIntent(context: Context, bookId: Long, noteId: Long): PendingIntent {
        if (BuildConfig.LOG_DEBUG) LogUtils.d(TAG, bookId, noteId)

        val intent = Intent.makeRestartActivityTask(ComponentName(context, MainActivity::class.java))

        intent.putExtra(AppIntent.EXTRA_BOOK_ID, bookId)
        intent.putExtra(AppIntent.EXTRA_NOTE_ID, noteId)

        return PendingIntent.getActivity(
            context,
            noteId.toInt(),
            intent,
            immutable(PendingIntent.FLAG_UPDATE_CURRENT))
    }

    fun keepScreenOnToggle(activity: Activity?, item: MenuItem): AlertDialog? {
        activity ?: return null

        if (!isKeepScreenOn(activity)) {
            return MaterialAlertDialogBuilder(activity)
                .setTitle(R.string.keep_screen_on)
                .setMessage(R.string.keep_screen_on_desc)
                .setPositiveButton(android.R.string.ok) { dialog, _ ->
                    keepScreenOnSet(activity)
                    item.isChecked = true
                    dialog.dismiss()
                }
                .setNegativeButton(android.R.string.cancel) { dialog, _ ->
                    dialog.dismiss()
                }
                .show()
        } else {
            keepScreenOnClear(activity)
            item.isChecked = false
            return null
        }
    }

    fun keepScreenOnUpdateMenuItem(activity: Activity?, menu: Menu) {
        val item = menu.findItem(R.id.keep_screen_on)
        if (activity != null && item != null) {
            if (AppPreferences.keepScreenOnMenuItem(activity)) {
                item.isChecked = isKeepScreenOn(activity)
            } else {
                menu.removeItem(item.itemId)
            }
        }
    }

    fun keepScreenOnClear(activity: Activity?) {
        activity?.window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun keepScreenOnSet(activity: Activity) {
        activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun isKeepScreenOn(activity: Activity): Boolean {
        val flags = activity.window.attributes.flags
        return flags and WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON != 0

    }

    fun distributeToolbarItems(activity: Activity?, toolbar: Toolbar) {
        if (activity != null) {
            val display = activity.windowManager.defaultDisplay

            val metrics = DisplayMetrics().also {
                display.getMetrics(it)
            }

            val screenWidth = metrics.widthPixels

            for (i in 0 until toolbar.childCount) {
                val childView = toolbar.getChildAt(i)

                if (childView is ViewGroup) {
                    val innerChildCount = childView.childCount

                    /*
                     * Use 1 less pixel for item width to avoid exception in tests:
                     * Caused by: java.lang.RuntimeException: Action will not be performed because the target view does not match one or more of the following constraints:
                     * at least 90 percent of the view's area is displayed to the user.
                     */
                    val itemWidth = screenWidth / innerChildCount - 1

                    for (j in 0 until innerChildCount) {
                        val grandChild = childView.getChildAt(j)

                        if (grandChild is ActionMenuItemView) {
                            grandChild.layoutParams.width = itemWidth
                        }
                    }
                }
            }
        }
    }

    @JvmStatic
    fun immutable(flags: Int): Int {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            flags or PendingIntent.FLAG_IMMUTABLE
        } else {
            flags
        }
    }

    @JvmStatic
    fun mutable(flags: Int): Int {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            flags or PendingIntent.FLAG_MUTABLE
        } else {
            flags
        }
    }
}

/** Don't fit decor to system windows only in portrait mode when the bottom toolbar is visible. */
fun Activity.setDecorFitsSystemWindowsForBottomToolbar(visibility: Int) {
    val orientation = resources.configuration.orientation
    if (visibility == View.VISIBLE && orientation == Configuration.ORIENTATION_PORTRAIT) {
        WindowCompat.setDecorFitsSystemWindows(window, false)
    } else {
        WindowCompat.setDecorFitsSystemWindows(window, true)
    }
}